src/librbd/mirror/snapshot/PrepareGroupImagesRequest.cc

```diff
@@ -0,0 +1,424 @@
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  std::vector<cls::rbd::GroupImageStatus> images;
+  if (r == 0) {
+    auto iter = m_outbl.cbegin();
+    r = cls_client::group_image_list_finish(&iter, &images);
+  }
+
+  if (r < 0) {
+    lderr(m_cct) << "error listing images in group: " << cpp_strerror(r)
+                 << dendl;
+    finish(r);
+    return;
+  }
+
+  auto image_count = images.size();
+  m_images.insert(m_images.end(), images.begin(), images.end());
+  if (image_count == MAX_RETURN) {
+    m_start_after = images.rbegin()->spec;
+    list_group_images();
+    return;
+  }
+
+  open_group_images();
+}
+
+template <typename I>
+void PrepareGroupImagesRequest<I>::open_group_images() {
+  ldout(m_cct, 10) << dendl;
+
+  auto ctx = librbd::util::create_context_callback<
+    PrepareGroupImagesRequest<I>,
+    &PrepareGroupImagesRequest<I>::handle_open_group_images>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  int r = 0;
+  size_t i = 0;
+  for ( ; i < m_images.size(); i++) {
+    auto &image = m_images[i];
+    librbd::IoCtx image_io_ctx;
+    r = librbd::util::create_ioctx(m_group_ioctx, "image",
+                                   image.spec.pool_id, {},
+                                   &image_io_ctx);
+    if (r < 0) {
+      m_ret_code = r;
+      break;
+    }
+
+    librbd::ImageCtx* image_ctx = new ImageCtx("", image.spec.image_id.c_str(),
+					       nullptr, image_io_ctx, false);
+
+    image_ctx->state->open(0, gather_ctx->new_sub());
+
+    m_image_ctxs->push_back(image_ctx);
+  }
+
+  gather_ctx->activate();
+}
+
+template <typename I>
+void PrepareGroupImagesRequest<I>::handle_open_group_images(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0 && m_ret_code == 0) {
+    m_ret_code = r;
+  }
+
+  if (m_ret_code < 0) {
+    lderr(m_cct) << "failed to open group images: " << cpp_strerror(m_ret_code)
+                 << dendl;
+    close_images();
+    return;
+  }
+
+  set_snap_metadata();
+}
+
+template <typename I>
+void PrepareGroupImagesRequest<I>::set_snap_metadata() {
+  ldout(m_cct, 10) << dendl;
+
+  m_group_snap->snapshot_namespace = cls::rbd::GroupSnapshotNamespaceMirror{
+    m_state, m_mirror_peer_uuids, {}, {}};
+
+  for (auto image_ctx: *m_image_ctxs) {
+    m_group_snap->snaps.emplace_back(image_ctx->md_ctx.get_id(), image_ctx->id,
+                                     CEPH_NOSNAP);
+  }
+
+  std::string group_header_oid = librbd::util::group_header_name(m_group_id);
+  // async version of cls_client::group_snap_set currently not available
+  int r = cls_client::group_snap_set(&m_group_ioctx, group_header_oid,
+                                     *m_group_snap);
+  if (r < 0) {
+    lderr(m_cct) << "failed to set group snapshot metadata: " << cpp_strerror(r)
+                 << dendl;
+    m_ret_code = r;
+    remove_snap_metadata();
+    return;
+  }
+
+  if (((m_internal_flags & SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE) == 0) &&
+      (!m_image_ctxs->empty())) {
```

Comment:
> Suggestion:
if(m_image_cxts->empty) { finish(0); return; }
Then proceed to the notify_quiesce/acquire_image_exclusive_locks calls.


Can m_image_ctxs be empty even after opening images was successful? May be we don't even have to check if m_image_ctxs is empty here?

I guess this would happen only when the group has no images, and hence m_images would be empty after listing group images. In that case I'd expect handle_list_group_images() to error out with ENOENT?
So maybe in `PrepareGroupImagesRequest<I>::handle_list_group_images()` we error out with `finish(ENOENT)`  if `images.empty()` ?

Reply:
m_image_ctxs can be empty if there are no images in the group. I don't think the list_group_images will error out - it should just return 0 images.

Reply:
Done